### PR TITLE
Reactive datastructures 100% test coverage

### PR DESCRIPTION
**Pull request by karimodm**

### CODE REVIEW

File: ds/reactive/clock_test.go [hunk 1]
```diff
@@ -1,22 +1,58 @@
...
+		t.Errorf("Expected variable to be initialized, but it was nil")
+	}
+
+	if clk.shutdown == nil {
+		t.Errorf("Expected shutdown channel to be initialized, but it was nil")
+	}
+}
+
+// TestClockDefaultTime tests the default time set in the clock struct
+func TestClockDefaultTime(t *testing.T) {
```
Comment (muXxer) [inline ds/reactive/clock_test.go]: why not also call it initial time?

should we maybe check the diff of `time.Now` and the variable to be less than some milliseconds?
then you won't have the flakiness by measuring `initialTime` directly at the start of the next second, and you would have a higher precision than 1 second of the thing you actually want to check?

File: ds/reactive/clock_test.go [hunk 2]
```diff
@@ -1,22 +1,58 @@
...
 
 func TestClock(t *testing.T) {
-	tickCount := 0
+	var tickCount atomic.Int32
 
 	clock := NewClock(100 * time.Millisecond)
 	clock.OnTick(func(now time.Time) {
-		tickCount++
+		tickCount.Add(1)
 		require.Less(t, time.Now().Sub(now), 100*time.Millisecond)
```
Comment (muXxer) [inline ds/reactive/clock_test.go]: So we measure that the time in the tick is not older than 100ms?

Why not a smaller value here? If you really want to catch mistakes here, I would at least follow the Nyquist–Shannon sampling theorem and measure 2 times faster than your frequency, which would be at least 50ms or smaller.

File: ds/reactive/eviction_state_test.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	state := newEvictionState[int]()
+
+	// Test with a slot that should create a new event
+	event := state.EvictionEvent(1)
+	require.NotNil(t, event, "EvictionEvent should not return nil for a new slot")
+
+	// Test with a slot that should not create a new event
+	state.lastEvictedSlot.Set(1)
+	event = state.EvictionEvent(0)
+	require.Equal(t, evictedSlotEvent, event, "EvictionEvent should return the evictedSlotEvent for a slot lower than lastEvictedSlot")
```
Comment (muXxer) [inline ds/reactive/eviction_state_test.go:34]: huh? where is this variable coming from? 
am I blind? or is this chatti output? :D

File: ds/reactive/eviction_state_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// TestEvict tests the Evict method
+func TestEvict(t *testing.T) {
+	state := newEvictionState[int]()
+
+	// Setup a slot to be evicted
+	slotToEvict := 1
+	state.evictionEvents.Set(slotToEvict, NewEvent())
```
Comment (muXxer) [inline ds/reactive/eviction_state_test.go:43]: mhh should we maybe also check in the same test that the event gets fired?

File: ds/reactive/eviction_state_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	_, exists := state.evictionEvents.Get(slotToEvict)
+	require.False(t, exists, "Evicted slot should no longer exist in evictionEvents")
+}
+
+// TestEvictPrivate tests the private evict method
+func TestEvictPrivate(t *testing.T) {
+	state := newEvictionState[int]()
+
+	// Test with a slot less than the current lastEvictedSlotIndex
+	state.lastEvictedSlot.Set(2) // Set the last evicted slot to 2
```
Comment (muXxer) [inline ds/reactive/eviction_state_test.go]: weird, we never set events that could be evicted in the first place, so how do we even test if the code works correctly here?

File: ds/reactive/eviction_state_test.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+	// Test with a slot greater than the current lastEvictedSlotIndex
+	slotToEvict := 3
+	state.evictionEvents.Set(slotToEvict, NewEvent())
+	eventsToTrigger = state.evict(slotToEvict)
+	require.Len(t, eventsToTrigger, 1, "evict should return one event to trigger when slot is greater than lastEvictedSlotIndex")
+}
+
+
+// Note: Additional test cases may be added to cover more scenarios based on your application's needs and logic.
```
Comment (muXxer) [inline ds/reactive/eviction_state_test.go]: lel :trollface:

File: ds/reactive/set_test.go [hunk 7]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+	s := newSet[int]()
+	otherSet := newSet[int](2, 3)
+	addedElements := s.AddAll(otherSet.ReadOnly())
+
+	require.Equal(t, 2, addedElements.Size())
+	require.Equal(t, 2, s.Size())
+
+	// Test adding duplicate elements
+	addedElements = s.AddAll(otherSet.ReadOnly())
+	require.Equal(t, 0, addedElements.Size())
```
Comment (muXxer) [inline ds/reactive/set_test.go:109]: we could also check that the `s.Size()` is still 2

File: ds/reactive/set_test.go [hunk 8]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+
+func TestSet_Delete(t *testing.T) {
+	s := newSet[int](1, 2, 3)
+	deleted := s.Delete(2)
+	require.True(t, deleted)
+	require.Equal(t, 2, s.Size())
+
+	// Test deleting a non-existent element
+	deleted = s.Delete(4)
+	require.False(t, deleted)
```
Comment (muXxer) [inline ds/reactive/set_test.go:121]: we could also check that the s.Size() is still 2

File: ds/reactive/set_test.go [hunk 9]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+	s := newSet[int](1, 2, 3, 4)
+	otherSet := newSet[int](2, 3)
+	deletedElements := s.DeleteAll(otherSet.ReadOnly())
+
+	require.Equal(t, 2, deletedElements.Size())
+	require.Equal(t, 2, s.Size())
+}
+
+// Similar tests for Apply, Compute, Replace, Decode, ReadOnly
+// Note: Mocking or stubbing may be required for some methods
```
Comment (muXxer) [inline ds/reactive/set_test.go]: remove that note...

File: ds/reactive/set_test.go [hunk 10]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+	result := mainSet.SubtractReactive(subtractSet)
+
+	// Initially, 2 should be subtracted from mainSet
+	require.False(t, result.Has(2))
+	require.True(t, result.Has(1))
+	require.True(t, result.Has(3))
+
+	// Update subtractSet and check if result set is updated
+	subtractSet.Add(3)
+	require.False(t, result.Has(3)) // Now 3 should also be subtracted
```
Comment (muXxer) [inline ds/reactive/set_test.go:162]: we could also check that the mainSet.Size() is 1

File: ds/reactive/set_test.go [hunk 11]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+		func(element int) (teardown func()) {
+			setupCalled = true
+			return func() {
+				teardownCalled = true
+			}
+		},
+	)
+	defer teardown()
+
+	rSet.Add(4) // Trigger setup
```
Comment (muXxer) [inline ds/reactive/set_test.go:185]: Mhh is there something like duplicates? Should we check that it doesn't call setup and theirdown for a duplicate?

File: ds/reactive/set_test.go [hunk 12]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+
+// TestDerivedSet_inheritMutations tests the inheritMutations method
+func TestDerivedSet_inheritMutations(t *testing.T) {
+	dSet := newDerivedSet[int]()
+	mutations := ds.NewSetMutations[int]().WithAddedElements(newSet(1, 2))
+
+	// Apply mutations
+	appliedMutations := dSet.inheritMutations(mutations)
+	require.True(t, appliedMutations.AddedElements().Has(1))
+	require.True(t, appliedMutations.AddedElements().Has(2))
```
Comment (muXxer) [inline ds/reactive/set_test.go:202]: Maybe an additional check for the dSet size?

File: ds/reactive/set_test.go [hunk 13]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+// TestDerivedSet_applyInheritedMutations tests the applyInheritedMutations method
+func TestDerivedSet_applyInheritedMutations(t *testing.T) {
+	dSet := newDerivedSet[int]()
+	mutations := ds.NewSetMutations[int]().WithAddedElements(newSet(1, 2))
+
+	// Apply mutations
+	inheritedMutations, triggerID, _ := dSet.applyInheritedMutations(mutations)
+	require.True(t, inheritedMutations.AddedElements().Has(1))
+	require.True(t, inheritedMutations.AddedElements().Has(2))
+	require.NotNil(t, triggerID)
```
Comment (muXxer) [inline ds/reactive/set_test.go:218]: Maybe an additional check for the dSet size?

File: ds/reactive/set_test.go [hunk 14]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+	}
+
+	// Apply the mutation
+	appliedMutations := s.Compute(mutationFactory)
+
+	// Check if the mutations are applied correctly
+	require.False(t, s.Has(2), "Element 2 should be deleted")
+	require.True(t, s.Has(4), "Element 4 should be added")
+	require.True(t, appliedMutations.DeletedElements().Has(2))
+	require.True(t, appliedMutations.AddedElements().Has(4))
```
Comment (muXxer) [inline ds/reactive/set_test.go:243]: Maybe an additional check for the s.Size()?

File: ds/reactive/set_test.go [hunk 15]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+	require.True(t, teardownCalled)
+}
+
+// TestDerivedSet_inheritMutations tests the inheritMutations method
+func TestDerivedSet_inheritMutations(t *testing.T) {
+	dSet := newDerivedSet[int]()
+	mutations := ds.NewSetMutations[int]().WithAddedElements(newSet(1, 2))
+
+	// Apply mutations
+	appliedMutations := dSet.inheritMutations(mutations)
```
Comment (muXxer) [inline ds/reactive/set_test.go:200]: I think it would be nice to have a duplicate in the inherited mutations?

File: ds/reactive/set_test.go [hunk 16]
```diff
@@ -75,3 +77,283 @@ func TestSubtract(t *testing.T) {
...
+}
+
+func TestSet_Encoding(t *testing.T) {
+	serix.DefaultAPI.RegisterTypeSettings("", serix.TypeSettings{}.WithLengthPrefixType(serix.LengthPrefixTypeAsByte))
+
+	testSet := newSet[string]()
+	for i := 0; i < 3; i++ {
+		testSet.Add(fmt.Sprintf("item%d", i))
+	}
+	bytes, err := testSet.Encode()
```
Comment (muXxer) [inline ds/reactive/set_test.go:362]: mhh should the interface of the reactive `Set` maybe get an API passed to Encode and Decode?

File: ds/reactive/variable_test.go [hunk 17]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+
+// TestVariable_InheritFrom tests the InheritFrom method of the variable type
+func TestVariable_InheritFrom(t *testing.T) {
+	parentVar := newVariable[int]().Init(10)
+	childVar := newVariable[int]()
+
+	unsubscribe := childVar.InheritFrom(parentVar)
+	defer unsubscribe()
+
+	parentVar.Set(20)
```
Comment (muXxer) [inline ds/reactive/variable_test.go:155]: We should also test for the 10 in the childVar before the Set(20).

File: ds/reactive/variable_test.go [hunk 18]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+func TestVariable_WithValue(t *testing.T) {
+	rVar := newVariable[int]()
+	rVar.Set(10)
+
+	setupCalled := false
+	teardownCalled := false
+
+	teardown := rVar.WithValue(
+		func(value int) (teardown func()) {
+			setupCalled = true
```
Comment (muXxer) [inline ds/reactive/variable_test.go]: should we check the value it got in the setup?

File: ds/reactive/variable_test.go [hunk 19]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+			return func() { teardownCalled = true }
+		},
+	)
+	defer teardown()
+
+	rVar.Set(20)
+	require.True(t, setupCalled)
+	require.True(t, teardownCalled)
```
Comment (muXxer) [inline ds/reactive/variable_test.go]: Why is the teardown here already called?

File: ds/reactive/variable_test.go [hunk 20]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+
+// TestVariable_WithNonEmptyValue tests the WithNonEmptyValue method of the variable type
+func TestVariable_WithNonEmptyValue(t *testing.T) {
+	rVar := newVariable[int]()
+	rVar.Set(10)
+	setupCalled := false
+
+	teardown := rVar.WithNonEmptyValue(
+		func(value int) (teardown func()) {
+			setupCalled = true
```
Comment (muXxer) [inline ds/reactive/variable_test.go]: Why don't we check for the teardown here as in the test before?
And we should check for the value as well.

File: ds/reactive/variable_test.go [hunk 21]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+}
+
+// TestVariable_OnUpdateOnce tests the OnUpdateOnce method of the variable type
+func TestVariable_OnUpdateOnce(t *testing.T) {
+	rVar := newVariable[int]()
+	rVar.Set(10)
+	callbackCalled := false
+
+	unsubscribe := rVar.OnUpdateOnce(func(oldValue, newValue int) {
+		callbackCalled = true
```
Comment (muXxer) [inline ds/reactive/variable_test.go]: Should be a counter, and we should set several values, to check if it is only called once.

File: ds/reactive/variable_test.go [hunk 22]
```diff
@@ -135,3 +135,121 @@ func TestOnUpdateOnce(t *testing.T) {
...
+	// Define a condition that is not met
+	condition := func(oldVal, newVal int) bool {
+		return newVal > 5 // Only trigger if the new value is greater than 5
+	}
+
+	// Register the callback with the condition
+	unsubscribe := rv.OnUpdateOnce(func(o, n int) {
+		oldValue = o
+		newValue = n
+		conditionNotMet = false
```
Comment (muXxer) [inline ds/reactive/variable_test.go]: Should be a counter, and we should set several values, to check if it is only called once or not.

**Review by muXxer — CHANGES_REQUESTED** [review]

File: ds/reactive/eviction_state_test.go (hunk 3 above)
Comment (karimodm) [reply ds/reactive/eviction_state_test.go:34]: It is defined in the eviction_state_impl as an event that already fired:
```
var evictedSlotEvent = func() Event {
	e := NewEvent()
	e.Trigger()

	return e
}()
```

File: ds/reactive/eviction_state_test.go (hunk 5 above)
Comment (karimodm) [reply ds/reactive/eviction_state_test.go]: The eviction state is just a map of events that get evicted in a rolling window fashion by firing the evicted events.

File: ds/reactive/eviction_state_test.go (hunk 4 above)
Comment (karimodm) [reply ds/reactive/eviction_state_test.go:43]: Checked in the other test already.

File: ds/reactive/set_test.go (hunk 16 above)
Comment (karimodm) [reply ds/reactive/set_test.go:362]: As the code is currently structured, this Encode is actually implemented by the OrderedMap!
We use the OrderedMap in way too many places to change that API and all the APIs that depend on it.

File: ds/reactive/set_test.go (hunk 15 above)
Comment (karimodm) [reply ds/reactive/set_test.go:200]: Added the element "2" twice: the set already excludes the repeated element though.

File: ds/reactive/set_test.go (hunk 10 above)
Comment (karimodm) [reply ds/reactive/set_test.go:162]: The mainSet size will stay as 3 throughout the test, but now I am checking that the result size changes reactively.

File: ds/reactive/variable_test.go (hunk 19 above)
Comment (karimodm) [reply ds/reactive/variable_test.go]: Before this Set there was no value to teardown, after this Set it is called over the value 10.

**Review by muXxer — APPROVED** [review]

**Pull request merged**